File: day_9/9.py
from typing import List, Literal


def move_train(
    board: List[str], mov: Literal["U", "D", "R", "L"]
) -> Literal["none", "crash", "eat"]:
    result = {"·": "none", "*": "eat", "o": "crash"}
    for i, row in enumerate(board):
        if "@" in row:
            train_y = i
            train_x = row.index("@")
            break
    next_move = {
        "U": [train_y - 1, train_x],
        "D": [train_y + 1, train_x],
        "L": [train_y, train_x - 1],
        "R": [train_y, train_x + 1],
    }
    next_y, next_x = next_move[mov]
    if next_y < 0 or next_x < 0 or next_y >= len(board) or next_x >= len(board[next_y]):
        return result["o"]
    return result[board[next_y][next_x]]


print(move_train(["·····", "*····", "@····", "o····", "o····"], "U"))
print(move_train(["·····", "*····", "@····", "o····", "o····"], "D"))
print(move_train(["·····", "*····", "@····", "o····", "o····"], "L"))
print(move_train(["·····", "*····", "@····", "o····", "o····"], "R"))

File: day_9/9.ts
type Space = "·" | "@" | "*" | "o";
type Board = string[];
type Movement = "U" | "D" | "R" | "L";
type Result = "none" | "crash" | "eat";

function moveTrainTS(board: Board, mov: Movement): Result {
  // Code here
  const result: Record<string, Result> = {
    "·": "none",
    "*": "eat",
    o: "crash",
  };
  const trainY: number = board.findIndex((row) => row.includes("@"));
  const trainX: number = board[trainY].indexOf("@");
  const nextMove: Record<Movement, [number, number]> = {
    U: [trainY - 1, trainX],
    D: [trainY + 1, trainX],
    L: [trainY, trainX - 1],
    R: [trainY, trainX + 1],
  };
  const [nextY, nextX] = nextMove[mov];
  if (
    nextY < 0 ||
    nextX < 0 ||
    nextY >= board.length ||
    nextX >= board[nextY].length
  ) {
    return result["o"];
  }
  return result[board[nextY][nextX]];
}

console.log(moveTrainTS(["·····", "*····", "@····", "o····", "o····"], "U"));
console.log(moveTrainTS(["·····", "*····", "@····", "o····", "o····"], "D"));
console.log(moveTrainTS(["·····", "*····", "@····", "o····", "o····"], "L"));
console.log(moveTrainTS(["·····", "*····", "@····", "o····", "o····"], "R"));
